flow/src/main/java/com/sequenceiq/flow/core/Flow2Handler.java

```diff
@@ -71,14 +75,37 @@
     @Inject
     private ApplicationFlowInformation applicationFlowInformation;
 
+    @Inject
+    private Tracer tracer;
+
     @Override
     public void accept(Event<? extends Payload> event) {
         String key = (String) event.getKey();
         Payload payload = event.getData();
         String flowId = getFlowId(event);
         String flowChainId = getFlowChainId(event);
         String flowTriggerUserCrn = getFlowTriggerUserCrn(event);
-        FlowParameters flowParameters = new FlowParameters(flowId, flowTriggerUserCrn);
+        Span activeSpan = tracer.activeSpan();
+        SpanContext spanContext = event.getHeaders().get(FlowConstants.SPAN_CONTEXT);
+        String operationName = event.getKey().toString();
+        if (TracingUtil.isActiveSpanReusable(activeSpan, spanContext, operationName)) {
+            doAccept(event, key, payload, flowId, flowChainId, flowTriggerUserCrn, spanContext);
+        } else {
+            Span span = TracingUtil.getSpan(tracer, operationName, spanContext, flowId, flowChainId, flowTriggerUserCrn);
+            if (spanContext == null) {
+                spanContext = span.context();
+            }
+            try (Scope scope = tracer.activateSpan(span)) {
+                doAccept(event, key, payload, flowId, flowChainId, flowTriggerUserCrn, spanContext);
+            } finally {
+                span.finish();
+            }
+        }
+    }
+
+    private void doAccept(Event<? extends Payload> event, String key, Payload payload, String flowId, String flowChainId,
```

Comment:
similar here, can we come up with a meaningful name? like `acceptWithSpanContext` or whatever

Reply:
I think doAccept is better, this is a common java terminology for calling a private method that does the actual think, like accepting in this case.